### ISSUE TITLE
Download and play Widevine protected content with DASH manifest

### ISSUE THREAD

**Issue by NVFedorov** [google/ExoPlayer#7202]
Hello,
I'm trying to modify your main example to make it able to play offline  a Widevine protected content. I managed to download licenses, but it's not clear for me how to download the video itself.
In your example we can play a protected content, but can not download it, because all the tracks in track selection dialog are disabled. I found out that it comes from TrackSelectionView.java updateViews(). There is a check if a track is supported. How the same track can not be supported for online playback and be not supported for offline?

Also I noticed that the codecs from DASH manifests are basically the same for not protected content:

      {
        "name": "WV: Clear SD (MP4,H264)",
        "uri": "https://storage.googleapis.com/wvmedia/clear/h264/tears/tears_sd.mpd"
      }

And for the protected content:

      {
        "name": "WV: Secure SD (cenc,MP4,H264) TEST",
        "uri": "https://storage.googleapis.com/wvmedia/cenc/h264/tears/tears_sd.mpd",
        "drm_scheme": "widevine",
        "drm_license_url": "https://proxy.uat.widevine.com/proxy?provider=widevine_test"
      }

Basically what I have at the end of the day:
I downloaded a license, got rid of checks for unsupported downloads and downloaded a video. When I'm trying to play it I'm getting "The content has audio tracks not playable by this device". But the codecs are the same. What I'm doing wrong?

Thanks in advance.

**Comment by AquilesCanta** [comment]
One possiblity is that you are not setting the drmSessionManager on the MediaSource that is playing the content. You'll observe a NO_UNSUPPORTED_DRM message in the track selection log. If you provide a bugreport we can confirm this for you.

**Comment by ojw28** [comment]
The problem is most likely that you're not passing the `DrmSessionManager` to `DownloadHelper` when setting up the download.

If you're modifying the demo app, you'll need to change it to use the `DownloadHelper.forXXX` methods that take a `DrmSessionManager`. This needs to be done [here](https://github.com/google/ExoPlayer/blob/r2.11.3/demos/main/src/main/java/com/google/android/exoplayer2/demo/DownloadTracker.java#L129).

**Comment by NVFedorov** [comment]
Yes, I made it work, but only using deprecated methods. The new exo player version didn't introduce methods to pass drmSessionManager to ExoPlayer Builder. 
To enable the track selection I created own method (copied it from library):

    private DownloadHelper getDownloadHelperForDash(DrmSessionManager drmSessionManager, Uri uri, RenderersFactory renderersFactory){
    DownloadHelper helper = new DownloadHelper(
        DownloadRequest.TYPE_DASH,
        uri,
        /* cacheKey= */ null,
        createMediaSourceInternal(
            DASH_FACTORY_CONSTRUCTOR,
            uri,
            dataSourceFactory,
            drmSessionManager,
            /* streamKeys= */ null),
        trackSelectorParameters,
        getRendererCapabilities(renderersFactory, drmSessionManager)
    );

    return helper;
  }

I only needed DASH and didn't try to do something with SS and HLS. So I changed getDownloadHelper signature:

    private DownloadHelper getDownloadHelper(
      DrmSessionManager drmSessionManager, Uri uri, String extension, RenderersFactory renderersFactory)

And I'm passing the drmSessionManager all the way from SampleChooserActivity.onSampleDownloadButtonClicked where I'm using the method I created in DemoApplication: 

    public DrmSessionManager<ExoMediaCrypto> buildOfflineDrmSessionManager(
        Uri uri,
        Sample.DrmInfo drmInfo) throws IOException, InterruptedException {
    
        if (Util.SDK_INT < 18) {
        return null;
        }

        HttpDataSource.Factory httpDataSourceFactory = this.buildHttpDataSourceFactory();

        MediaDrmCallback mediaDrmCallback =
            createMediaDrmCallback(drmInfo.drmLicenseUrl, drmInfo.drmKeyRequestProperties);

        DefaultDrmSessionManager<ExoMediaCrypto> drmSessionManager =
            new DefaultDrmSessionManager.Builder()
                .setUuidAndExoMediaDrmProvider(drmInfo.drmScheme, FrameworkMediaDrm.DEFAULT_PROVIDER)
                .setMultiSession(drmInfo.drmMultiSession)
                .build(mediaDrmCallback);

        String spKey = this.userAgent + "_" +
            uri.toString().replace(".", "_");
        String prefKey = String.valueOf(spKey.replace("/", "_").hashCode());

        String offlineAssetKeyIdStr = this
            .getSharedPreferences(prefKey, MODE_PRIVATE)
            .getString(KEY_OFFLINE_OFFSET_ID, DemoApplication.EMPTY);
        Pair<Long, Long> remainingSecPair = new Pair<>(0L, 0L);

        OfflineLicenseHelper offlineLicenseHelper = new OfflineLicenseHelper<>(
            C.WIDEVINE_UUID,
            FrameworkMediaDrm.DEFAULT_PROVIDER,
            mediaDrmCallback,
            createOptionalKeyRequestParameters(drmInfo.drmKeyRequestProperties));

        byte[] offlineAssetKeyId = Base64.decode(offlineAssetKeyIdStr, Base64.DEFAULT);
        if (!DemoApplication.EMPTY.equals(offlineAssetKeyIdStr)) {
        remainingSecPair = offlineLicenseHelper.getLicenseDurationRemainingSec(offlineAssetKeyId);
        }

        Log.e(TAG,
            " License remaining Play time : " + remainingSecPair.first + ", Purchase time : "
                + remainingSecPair.second);
        if (DemoApplication.EMPTY.equals(offlineAssetKeyIdStr) || (remainingSecPair.first == 0
            || remainingSecPair.second == 0)) {
        Log.e(TAG, "will start download now");
        DataSource dataSource = httpDataSourceFactory.createDataSource();
        DashManifest dashManifest = DashUtil.loadManifest(dataSource, uri);
        DrmInitData drmInitData = DashUtil.loadDrmInitData(dataSource, dashManifest.getPeriod(0));
        offlineAssetKeyId = offlineLicenseHelper.downloadLicense(drmInitData);
        Pair<Long, Long> p = offlineLicenseHelper.getLicenseDurationRemainingSec(offlineAssetKeyId);
        Log.e(TAG, "download done : " + p.toString());

        SharedPreferences sharedPreferences = this.getSharedPreferences(prefKey, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(KEY_OFFLINE_OFFSET_ID,
            Base64.encodeToString(offlineAssetKeyId, Base64.DEFAULT));
        editor.commit();
        }

        drmSessionManager.setMode(DefaultDrmSessionManager.MODE_QUERY, offlineAssetKeyId);
        return drmSessionManager;
    }

    private static Map<String, String> createOptionalKeyRequestParameters(
        String[] keyRequestPropertiesArray) {
        Map<String, String> result = new HashMap<>();
        if (keyRequestPropertiesArray != null) {
        for (int i = 0; i < keyRequestPropertiesArray.length - 1; i += 2) {
            result.put(keyRequestPropertiesArray[i], keyRequestPropertiesArray[i + 1]);
        }
        }

        return result;
    }

    public HttpMediaDrmCallback createMediaDrmCallback(
        String licenseUrl, String[] keyRequestPropertiesArray) {
        HttpDataSource.Factory licenseDataSourceFactory = buildHttpDataSourceFactory();
        HttpMediaDrmCallback drmCallback = new HttpMediaDrmCallback(licenseUrl, licenseDataSourceFactory);
        if (keyRequestPropertiesArray != null) {
            for (int i = 0; i < keyRequestPropertiesArray.length - 1; i += 2) {
                drmCallback.setKeyRequestProperty(keyRequestPropertiesArray[i],
                    keyRequestPropertiesArray[i + 1]);
            }
        }
        return drmCallback;
    }


To play the downloaded content I had to pass drmSessionManager to the player:

    Sample intentAsSample = Sample.createFromIntent(intent);
      try {
        DrmSessionManager drmSessionManager = ((DemoApplication)getApplication())
            .buildOfflineDrmSessionManager(((Sample.SampleWithUri) intentAsSample).uri, ((Sample.SampleWithUri) intentAsSample).drmInfo);

        player = ExoPlayerFactory.newSimpleInstance(
            this,
            renderersFactory,
            trackSelector,
            new DefaultLoadControl(),
            drmSessionManager,
            DefaultBandwidthMeter.getSingletonInstance(this));

      } catch (Exception e) {
        e.printStackTrace();

        player =
            new SimpleExoPlayer.Builder(/* context= */ this, renderersFactory)
                .setTrackSelector(trackSelector)
                .build();
      }

Obviously this is not a final solution and I beg you a pardon for this really smelly code, but at least it's working. 

P.S.: as many others I kindly ask you to create some demo for Offline DRM Playback. At least for me it took like a week to come up with this "solution". I hope it will help somebody.

**Comment by ojw28** [comment]
The deprecated method you're using has instructions on what to use instead ([see here](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/ExoPlayerFactory.html#newSimpleInstance-android.content.Context-com.google.android.exoplayer2.RenderersFactory-com.google.android.exoplayer2.trackselection.TrackSelector-com.google.android.exoplayer2.LoadControl-com.google.android.exoplayer2.drm.DrmSessionManager-com.google.android.exoplayer2.analytics.AnalyticsCollector-)). Are you saying that they don't work?

**Comment by NVFedorov** [comment]
I didn't see that it should be done through media source factory, thank you
